syfertext/pipeline/simple_tagger.py

```diff
@@ -0,0 +1,119 @@
+        """Initialize the TokenTagger object.
+
+           Args:
+               attribute (str): The name of the attribute that will hold the tag.
+                   this attribute will be accessible through the attribute
+                   `._` of Token objects. Example `token_object._.<attribute>
+               lookups (set, list or dict): If of type `list` of `set`, it should contain
+                   the tokens that are to be searched for and tagged in the Doc 
+                   object's text. Example: ['the', 'myself', ...]
+                   If of type `dict`, the keys should be the tokens texts to be
+                   tagged, and values should hold a single tag for each such token.
+                   Example: tagging stop words {'the': True, 'myself' : True}.
+               tag (object, optional): If `lookups` is of type `list`, then this
+                   will be the tag assigned to all matched tokens. It will be 
+                   ignored if `lookups` if of type `dict`.
+               default_tag: (object, optional): The default tag to be assigned
+                   in case the token text maches no entry in `lookups`.
+               case_sensitive: (bool, optional): If set to True, then matching
+                   token texts to `lookups` will become case sensitive.
+                   Defaults to True.
+        """
+
+        self.attribute = attribute
+
+        # Desensitize tokens in lookups if `case_sensitive` is False
+        if case_sensitive:
+            self.lookups = lookups
+        else:
+            self.lookups = self._desensitize_lookups(lookups)
+
+        # If `lookups` is a `list`, convert it to a `set`
+        self.lookups = (
+            set(self.lookups) if isinstance(self.lookups, list) else self.lookups
+        )
+
+        self.case_sensitive = case_sensitive
+        self.tag = tag
+        self.default_tag = default_tag
+
+    def __call__(self, doc: Doc):
+
+        # Start tagging
+        for token in doc:
+
+            # Get the tag
+            tag = self._get_tag(token)
+
+            # Set the attribute of each matched token to the tag
+            token.set_attribute(name=self.attribute, value=tag)
+
+    def _desensitize_lookups(self, lookups: Union[dict, list, set]):
+        """Converts every token in `self.lookups` to lower case to enable
+           case in-sensitive matching
+
+           Args:
+               lookups (set, list or dict): Check out the docstring of `__init__()`.
+
+           Returns:
+               A transformed version  of `lookup` where all token texts are in 
+               lower case.
+ 
+        """
+
+        # Replace dict keys with lower-case versions
+        if isinstance(lookups, dict):
+            return {token.lower(): lookups[token] for token in lookups}
+
+        # Convert all list or set elements to lower case
+        if isinstance(lookups, list) or isinstance(lookups, set):
+            return {token.lower() for token in lookups}
```

Comment:
I am just thinking about what the behavior should be when the type of `lookups` is not dict, list or set. Right now, this function would return `None` in those cases and the initialization would proceed without an issue. But when a user tries to tag something I believe that `_get_tag()` would raise an error saying that the variable `tag` is not defined (since it is only created inside the if/elif statements).

We could raise an error right away in `__init__` and force that `lookups` is not dict, list or set or we could just add an `else` statement like:

```
else:
    tag = self.default_tag
```

What are your thoughts @AlanAboudib?

Reply:
This is a good remark. However, for the moment, I suppose that it is the responsibility of SyferText user to respect the documentation and argument types.  

For security concerns, we could introduce a class of utility methods/decorators to check the type/format of each argument for ALL SyferText methods. To be discussed in the upcoming meeting.